### DISCUSSION TITLE
Proposal: Ternary comparison operator

### DISCUSSION THREAD

**Discussion by CyrusNajmabadi** [dotnet/csharplang#8643] [Language Ideas]
# Ternary comparison operator

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

- Tracking issue here: https://github.com/dotnet/csharplang/issues/8861

## Summary
This would allow users to write a simplified `x < y < z` test as shorthand for `x < y && y < z`.  

## Detailed design
This code is already parseable today and already has meaning.  Indeed, here is a (albeit pathological) case where this would compile today:

```C#
using System;
public static class Program {
    public static void Main() {
        var (x, z) = (new C(), new C());
        Console.WriteLine(x < 5 < z);
    }
}

public struct C
{
    public static bool operator <(C x, C y) => true;
    public static bool operator >(C x, C y) => true;
    
    public static C operator <(C x, int y) => x;
    public static C operator >(C x, int y) => x;
}
```

In order for this type of code to compile today, you'd need to do very strange operator overloading which would not be expected in practice.  However, because of back compat, we would likely have to support this.

As such, when processing a binary expression of hte form `expr1 op expr2 op expr3` we would have to bind in the same fashion as today.  However, if that form failed to bind the operators successfully, we would now reinterpre the above as:

```expr1 op1 expr2 op2 expr3```, where `op1` and `op2` are one of `>`, `<`, `>=`, `<=` is reinterpretted as:

```
var __t1 = expr1;
var __t2 = expr2;

var r = false;
if (__t1 op1 __t2)
{
    var __t3 = expr3;
    if (__t2 op2 __t3)
        r = true;
}
```

This should match intuition here and would mean code executes (including order of evaluation and short-circuiting) as expected.  

## Drawbacks
Potential confusion over a piece of code potentially having two different meanings.  However, this is so unlikey as no real codebases should ever have been using `a < b < c`.  It just isn't a reasonable code pattern today, so no one uses it or expects it to work the way it does today.  Practically all users looking at this would expect it to have the semantics this proposal is suggesting.

**Comment by HaloFour** [comment]
It was noted that there is a NuGet package which uses an extension method and some overloaded operator chicanery to mimic this language feature: https://github.com/dotnet/roslyn/issues/136#issuecomment-71959890

However, if the compiler will only treat this expression as a range comparison if it wouldn't otherwise compile then I would imagine any existing code that might rely on this package would continue to compile and work as expected.  But it should probably be included as a part of the test suite to ensure no unintended breaks.

**Comment by CyrusNajmabadi** [comment]
Good to know.  This would fall under the portion of the design of: we will existing semantics and always interpret that way if it succeeds.  We only use the new semantics if it doesn't.

That example is also interesting in that it's trying to provide these semantics.  So a great part about this is that once we have this feature, you can stop using that lib :)

**Comment by YairHalberstadt** [comment]
> we would have to bind in the same fashion as today.

How would this be defined to avoid making further changes to binding not overly complex?

For example if C# adds increases the number of places implicit operators can be used in the future, or changes lookup, or improves overload resolution, we would need to make sure all those things run only after we check if it matches this new ternary comparison operator. This might force the binding code to become very contorted to allow this.

**Comment by CyrusNajmabadi** [comment]
honestly, i would say:  given an tree of the form

```
      op1
     /  \
expr1    op2
        /  \
   expr2    expr3
```

We have the existing rules.  We then have a clause htat says that if that produces an error (which we use terminology for in lambda-resolution) then rebind with such-and-such expected semantic rewrite.    In general we discuss validity, and binding-time errors.  This would apply here.  If the normal interpretation results in an invalid (or binding-time-error), we try the new interpretation.

I'm not sure any of the cases mentioned so far necessary matter. Or, if they do, they are starting to get to the cornerest corner of all corners :)

**Comment by alrz** [comment]
Linking back to my comment here https://github.com/dotnet/csharplang/discussions/4106#discussioncomment-122571

Since you're going to deal with some semantic ambiguities anyways, the postfix pattern does seem to be a viable alternative, except that there will be syntax ambiguities instead (but not in any useful scenarios mentioned here).


Opened a new discussion: https://github.com/dotnet/csharplang/discussions/4110

**Comment by LRC-H** [comment]
很好

**Comment by huoyaoyuan** [comment]
Can there be a warning wave when bindinh to the old behavior?

**Comment by CyrusNajmabadi** [comment]
> Can there be a warning wave when bindinh to the old behavior?

I don't see any point in doing that

**Comment by hez2010** [comment]
There may be some ambiguities need to be resolved, such as `A < B < C > D > F`.

**Comment by AdamSpeight2008** [comment]
Wouldn't this potentially break compatibility?  Those operators are not guaranteed to represent comparisons. the user is free to give them different semantics.

**Comment by AdamSpeight2008** [comment]
This would have to be under a feature flag, so that the change in semantic doesn't change errors in previous version.

Can work just via the binder and lowering.
eg. `a < b <= c`
Requires all the operands `a b c` to be the same type;
and that that type have operator `<` equivalent to `Func< T, T, Bool>`.
and that that type have operator `<=` equivalent to `Func< T, T, Bool>`.
Then in lowering compiler generate `( (a < b) && (b <= c)`.


> It was noted that there is a NuGet package which uses an extension method and some overloaded operator chicanery to mimic this language feature: [dotnet/roslyn#136 (comment)](https://github.com/dotnet/roslyn/issues/136#issuecomment-71959890)
> 

I wrote that nuget package. Note it use `IComparable(Of T)` has a the check to see if the type has "comparison operators".
It also introduces 2 new class types `_0(Of T)` (the lifted type) and `_1(Of T)` to provide access to the second "comparison operator.

**Comment by canton7** [comment]
Compatibility concerns are fully addressed in the OP, no?

**Comment by CyrusNajmabadi** [comment]
> Wouldn't this potentially break compatibility?

No.  The specification can effectively be read to say:

If this has legal semantics under C# 9.  The use those semantics.  Othewise, try the new semantics.  So nothing can break.

**Comment by AdamSpeight2008** [comment]
OP doesn't mention if  `a b c` can or can not be different types.
My addition to the proposal, by restricting to only the same focuses it to the conventional comparison operations. `Func<Func<T, T, Bool>, T, Bool>` 
Which is currently an error.

**Comment by CyrusNajmabadi** [comment]
> I wrote that nuget package. 

Your package will retain its semantics.

**Comment by CyrusNajmabadi** [comment]
> OP doesn't mention if a b c can or different types.

It can operator on different types as per the rewrite rule i've specified.  The rule is agnostic to that.

**Comment by AdamSpeight2008** [comment]
@CyrusNajmabadi  Are you restricting to the "comparison" operator only if it returns `boolean`?
As if wasn't, it would cause that nuget package to have incorrect semantics.
```
lowerValue.__() <= value <= upper
' types produced during evaluation.
T.__()  --> _0(Of T) ' The "Lifted" type
_0(Of T) <= T --> _1(Of T)  ` provide the first operators (<, <=)
_1(Of T) <= T --> Boolean   ' provides the second operators (<. <=)
````

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Are you restricting to the "comparison" operator only if it returns boolean?

No

**Comment by CyrusNajmabadi** [comment]
> As if wasn't, it would cause that nuget package to have incorrect semantics.

Any existing semantics would be preserved.

the general intuition for the algorithm is:

1. try existing C# 9.0 semantics.  If that succeeds, those are the semantics to use.
2. otherwise, if that fails, interpret `a < b < c < d < e` as `a < b && b < c && c < d  && d < e` and try to bind.    if that succeeds, emit in that fashion.
3. otherwise, fail.

**Comment by NN---** [comment]
With C# 9 it can be written
`a is > 5 and < 10` which doesn’t have ambiguity and allows to mix and match comparisons.

**Comment by FaustVX** [comment]
@NN--- 
> With C# 9 it can be written
> `a is > 5 and < 10` which doesn’t have ambiguity and allows to mix and match comparisons.

No, because, Pattern Matching works only with constant, and with this proposal, it could works with any value (variables, properties, …)

**Comment by Daynvheur** [comment]
Little comment here (thanks to @alrz for the link) as I've opened a very similar discussion at #4980 , but with an `or` comparison.
Thus, question: Would this ternary comparison only work on `and` mode? if not, how would you express `A < C [or] A > B`?

**Comment by Xyncgas** [comment]
> Potential confusion over a piece of code potentially having two different meanings. 

We have <= and => today, tell me they are the same thing.

**Comment by The-Futurist** [comment]
> There may be some ambiguities need to be resolved, such as `A < B < C > D > F`.

What ambiguity is in that fragment? Creating the parse tree for that is not ambiguous so far as I can see.

**Comment by CyrusNajmabadi** [comment]
It would need appropriate lookahead to ensure that generics, variables, and comparisons were properly handled. It's likely not a full ambiguity, but a local one.

**Comment by 333fred** [comment]
> What ambiguity is in that fragment?

Is it `A<B<C>D> F` (a declaration of a variable `F` with type `A<B<C>D>` and a syntax error on a missing `,`) or a chained comparison operator?

**Comment by The-Futurist** [comment]
> > What ambiguity is in that fragment?
> 
> Is it `A<B<C>D> F` (a declaration of a variable `F` with type `A<B<C>D>` and a syntax error on a missing `,`) or a chained comparison operator?

Yes I see that when there's no context, but in the case of an `if` statement we have the context, the grammar rules of the `if`.

A declaration isn't legal (well I can't see it myself) inside the expression that's part of an `if`. The parser would only encounter this when parsing a conditional expression. C# already correctly distinguishes between generics and expressions.

C# seems to already parse this correctly too:

```cs
int A = 0;
int B = 0;
int C = 0;
int D = 0;
int F = 0;

if (A<B<C>D>F)
{
    ;
}
```

The diagnostic it reports is not about syntax but a semantic problem, the _types_ that are used either side of the operator `<`.

This too suggests that if the semantic checking were replaced, the expression interpreted as a chained compare, it would work.

**Comment by FaustVX** [comment]
> There may be some ambiguities need to be resolved, such as `A < B < C > D > F`.

As far as I can test, the compiler is already happy with that syntax
[sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYAMBYAUBgRjz1QGYACAUyggFtyBhAe1oAc8BvPcn8gEQCWAMyGUw1YLylSAvOUwAjTDG68AMpQDOmgCoALAIZRpJ8nMUFyAHivyVuKQHFxB4GP1HTPcwss3yBPZSAKIAjhAGADaaXiY+frboQepauoZQAPJgYRHRsuQa2h7GyOQ5UZrJPM6Uru7pWeV5cjV1YMXkpU2VeAC+xPgUGIycqjyoBABs5KgALOQAsgYCUAAUAJRj5FwOpgBuBmDkqwCCMOQAQucM53znwecAYutmx1CUAO6MG+fvXww/ch/b7rX6fEFg/6A4EA9brADcWykByOAGMWKxXidrJccQxyAA+fiE8iPRG7Hj9CkzIZTRgY8hMVhiVxMI5WVb4yKUITAa7kMACADmemAm2psiJzDYADpBCIxBJyVIyDM6dLMUyWcA2YTOeRubz+YKRWKkbwZFKMTLusreKqJtMNYzmWBWezOQzDXzGALhaLxV5LQaecA7eNaU6GVq3TqjgTPWwQ0bfSaA+bvETveGaWqo0mY+6zDJ9d7jf6zRKLVbZfLROIoGGtg71dHXUWAIQlrmh8umwOmYMam3hCo5lv5zXtuPFxOYsupisDuJZ0PjyP0gvT3VdufJn34tOVoOr3nrvObqfa3VWbv7vvpquZy9y4T1pXNjfOwszgl3heHkuGZmDWrAjrkmjno6l4ute7Ldl6vaLv2wHBtmn4Xt+27xghSYAX6KFPiB+7kr0QA===)

**Comment by CyrusNajmabadi** [comment]
> Yes I see that when there's no context, but in the case of an if statement we have the context, the grammar rules of the if.

Yes. This is an example of us addressing the potential issue. The situations brought up are there so we ensure we think through it and make sure the language is spec'ed such that it is not an issue and so that the compiler operates as expected with appropriate tests.

**Comment by theking2** [comment]
I really don't understand this. There is also ambiguity with another ternary operator `?:`. It was addressed there why is `op cmp op cmp op` different?
 Clearly there are ambiguities but with ?: we seem perfectly fine with those.

**Comment by CyrusNajmabadi** [comment]
> I really don't understand this. There is also ambiguity with another ternary operator ?:

What ambiguity are you referring to?

> Clearly there are ambiguities but with ?: we seem perfectly fine with those.

No one said we are not fine with ambiguities... As i said in the post above yours, we would just need to be cognizant so we can spec hte language out properly, and ensure the compiler does the right thing.

**Comment by theking2** [comment]
what does a? b? c: d: do and what does a <  b <  c < d do.  both ambiguous . () needed to fix this

**Comment by CyrusNajmabadi** [comment]
> `a? b? c: d: `

The code is incomplete and will be a failure to parse.  Did you miss a `e` at the end?  I'm not sure if there's an accidental omission or a purposeful one. 

Assuming it's accidental, and you meant: `a? b? c: d: e`, then there is no ambiguity here (syntactic or otherwise). 

Can you please clarify?

**Comment by CyrusNajmabadi** [comment]
> and what does `a < b < c < d` do

Currently it is syntactically legal, and may have semantic meaning in esoteric cases. This proposal discusses giving it legal meaning in the case where it is illegal today.

**Comment by Daynvheur** [comment]
Random concerns:
- What would be the interest (if any) of a non- short-circuited version of this? (Making sure that `a < b < c < d < e` do evaluate the `e` part.)
- Syntactic analysis simplifications: `a < 3 < 1` could be simplified as `a < 1`? (`a` required to be less than 3 and less than 1 is satisfied with just `a` required to be less than 1.)
- Warn about impossible conditions: `3 > a < 1`? (`a` required to be more than 3 and less than 1 can never be satisfied.)

**Comment by CyrusNajmabadi** [comment]
> What would be the interest (if any) of a non- short-circuited version of this? (Making sure that a < b < c < d < e do evaluate the e part.)

My thought here is that this is shorthand for `a < x && x < b`.  So it shoudl have the same shortcircuiting behavior.

> Syntactic analysis simplifications: a < 3 < 1 could be simplified as a < 1? (a required to be less than 3 and less than 1 is satisfied with just a required to be less than 1.)

Looks like the space for an optimization pass, or an analyzer message.  Not really in scope for the language itself.

> Warn about impossible conditions:

That seems reasonable.  We already do constant analysis and requisite flow control for similar htings today:

![image](https://user-images.githubusercontent.com/4564579/224769001-8c0f4a10-1d8d-4447-99c7-ac3617ccaf1c.png)

So this could similarly evaluate to always being false (if 'a' is an integer of course).

**Comment by MgSam** [comment]
C++ is also looking to add this; albeit with disallowing the potentially ambiguous forms like `a < b > c`. I think if C# adds this the ambiguous forms should also be disallowed; they add no benefit.

**Comment by elephant-koala** [comment]
What about this syntax instead:

```
if( x in 3f...5f ){

}
```

basically using a variant of range notation [3,5] or (3,5] or (3,5) etc. 
Perhaps we could write [3]...5 to indicate 3 is exluded from the range.  
Other syntax could be `x in 3f to [5f]`

One could even do joins like:

```
if( x in 3f...5f or 10f...25f){

}
```

Would work with switch statements too:

```
switch(x){
    case: 3f...5f:
    brealk;
    case: 10f...25f:
    break;
}
```
Just an idea!

**Reply by Daynvheur** [reply]
I may think that comparison with "in range" should be a different proposal, but a very valid one.

Maybe include collections in the mix, like `if (x in List<WeekDays>(){Monday, Tuesday, Sunday}) { .. }`, this would help aswel.

**Reply by CyrusNajmabadi** [reply]
we have another prposal that would just allow `[monday, tuesday, sunday].Contains(x)`.  Which is natural and composes over what the language and APIs already support.

**Reply by Daynvheur** [reply]
As here is not the place to argue around the legitimacy of said alternative collection containment checkings, I would just raise the concern of the diversity of .Net's syntaxes:
- The "native" way, used in standard operators `foreach (var x in Xs) { ../*returns contained state*/ }`
- The "Linq" way, used in query operations `Xs.Do(x => { .. })` (here "Do" would be "Contains")
- The "extension" way `static bool Do<T>(this T value, params T[] values); /**/ x.Do(Xs);` (here "Do" would be "In")

_It would have helped to link back to #316 (found from #9926's comments, which is one of numerous similar recent propositions), I may transfer this comment over there._

**Reply by CyrusNajmabadi** [reply]
> The "native" way, used in standard operators for (var x in Xs) { .. }

This is not a containment check. That's just iteration. :-)

And both of the other ways are just the same, just allowing alternate logic over default equality semantics.

**Reply by Daynvheur** [reply]
> That's just iteration. :-)

Yeah, it is a valid comment for the (too short) code extract I provided for `bool IsContained<T>(T x,T[] Xs){foreach(var a in Xs)if(x==a)return true;return false;}if (IsContained(x, Xs)){..}`, which boils down to the equivalent of Linq's `.Contains()`.
You could use this `IsContained(x, Xs)` algo for the range aswell, and everything could be fine as it is.

It's just a recurring demand which doesn't look out of (.Net's) place to read `if (x is in Xs) { .. }`, but here is not the place to discuss that proposition.

I wasn't able to comment on #316, because it was closed by the author in 2017, at a time where Ranges were still in discussion.
Looking at a lot of issues back-referencing #316 as the trust source, I'll fade away from here.

Edit: Here is a more extended example of things working already without extension methods (including Linq APIs)
```cs
using System;
if(IsContained('a',['c','a','t']))Console.WriteLine("yes"); //? if (a is in ['c', 'a', 't'])
if(IsContainedRange(1,0..3))Console.WriteLine("yes"); //? if (1 is in 0..3)
bool IsContained<T>(T x,T[] Xs){foreach(var a in Xs)if(x?.Equals(a)==true)return true;return false;}
bool IsContainedRange(int x,Range Xs)=>x>=Xs.Start.Value&&x<=Xs.End.Value;
```

**Comment by wherewhere** [comment]
`1 < 2 < 3` break the extension of `bool <(int, bool)`
```cs
Console.WriteLine(1 > 2 > -1);
static class A
{
    extension(bool)
    {
        public static bool operator >(bool left, int right)
        {
            return Unsafe.As<bool, int>(ref left) > right;
        }
        public static bool operator <(bool left, int right)
        {
            return Unsafe.As<bool, int>(ref left) < right;
        }
    }
}
```

[Playground](https://wherewhere.github.io/SharpScript/#code=K4Zwlgdg5gBAygTxAFwKYFsDcAoUlaIoYB0ASqgGYA2qAxsmAPYQ57TxJrpnAQPqpiAYUboADmBoAnOKikA3MLVQgcAbQCGIEBgBGVBAC4YAQW16DANTnhmACgBEABmIu3DgJQBdbCIghGGmIAdSkwNAAZSFQ7AEYYAD4YACZEmABaWI8cFA0GWhhaKi0QU2wAb2wYaphUAA80fyYIO11GQI8qmsqa3pgxYH0lGFz8mDbAmEYxOTzGKUTW9qoYGgpkABoYSGQYMKgAC2ROvu6u096pVGRgKQgYAFV-DQpBMwAeCaotnYS7K4oq0oxzS%2ByOOAuMAAvuc%2BgMhgVRsMvlMZlI5gt3ktJmtNts%2BHswIdjrDej1ITUrjc7o9nq9iB8vj8%2BH8AUD1h4YO9CcSIRcYb0YVCgA)

It will be better if we can write
```cs
bool A(int a, int b, int c) => a > b and < c;
```

**Reply by CyrusNajmabadi** [reply]
> 1 < 2 < 3 break the extension of bool <(int, bool)

Nothing breaks with the proposal. It is very clear on that.

**Comment by GeeLaw** [comment]
Moved from https://github.com/dotnet/csharplang/discussions/10150#discussioncomment-17351254

---

Re: chained comparison.

TLDR, two issues:

1. `(double)1 < (int)2 < (decimal)3` won't work, but should.
2. It could cause the middle operand to be twice-converted by user-defined implicit conversions, which is highly unusual for C# because no other current construct allows chained user-implicit conversions without any of them explicitly spelled out.

---

Let's look at this:

> Note: When each relational_expression node in a chain e₀ op₁ e₁ op₂ e₂ … opₙ eₙ is a chained relational comparison, the chain is equivalent in result to (e₀ op₁ e₁) && (e₁ op₂ e₂) && … && (eₙ₋₁ opₙ eₙ), with each eᵢ evaluated at most once. end note

This note is incorrect, because of

> Conversions on the shared middle operand. The isolated overload resolution of Y op B is applied against Y's classification as the right operand of X op' Y: that is, the conversion (if any) that §12.4.5 applied to Y for the inner link is already part of Y's compile-time classification at this point, and the resolution of Y op B begins from there. §12.4.5 as applied to Y op B may in turn select a conversion of its own on top of that classification, to bring Y to the operator's left-operand type. Each such conversion is applied at its own link's point of comparison; Y is still evaluated only once, and its single value flows through both links with the appropriate conversion applied at each.

... which causes the middle operand to be subject to up to 2 user-defined implicit conversions. Suppose

- `T1`, `T2`, `T3` are user-defined types,
- there are implicit conversions from `T3` to `T1` and from `T1` to `T2`, but no other among them,
- there are `bool T1.operator <(T1, T1)` and `bool T2.operator <(T2, T2)`, but no other except for symmetric overloads,
- `t1` is an expression (parenthesized, if we must) of type `T1`, and `t2`, `t3` similarly,

then `t1 < t3 < t2` compiles under the current proposal but `t1 < t3 && t3 < t2` does not.

```csharp
T1 t1 = default;
T2 t2 = default;
T3 t3 = default;

// Currently proposed meaning of `t1 < t3 < t2`:
T1 middle = t3;                 // Together,
Console.WriteLine(t1 < middle); // implicit conversion T3 -> T1.
Console.WriteLine(middle < t2); // Implicit conversion T1 -> T2.

// The note-suggested rewriting is wrong.
// Console.WriteLine(t1 < t3); // OK, implicit conversion.
// Console.WriteLine(t3 < t2); // No, cannot perform two user-defined conversions without one explicitly invoked.

struct T1
{
  public static implicit operator T1(T3 obj) { Console.WriteLine("T3 -> T1"); return default; }
  public static implicit operator T2(T1 obj) { Console.WriteLine("T1 -> T2"); return default; }
  public static bool operator <(T1 a, T1 b) { Console.WriteLine("T1 < T1"); return default; }
  public static bool operator >(T1 a, T1 b) { Console.WriteLine("T1 > T1"); return default; }
}
struct T2
{
  public static bool operator <(T2 a, T2 b) { Console.WriteLine("T2 < T2"); return default; }
  public static bool operator >(T2 a, T2 b) { Console.WriteLine("T2 > T2"); return default; }
}
struct T3
{
}
```

The choice to twice-convert the middle operand is weird. Why not define `X op1 Y op2 Z` as follows?

- If `X op1 Y` is classified as a value of type `bool` and overload resolution for `Y op2 Z` with `Y` being its original type (not already subject to implicit conversion during overload resolution of `X op1 Y`) chooses an operator returning `bool`, then the expression is a chained relational comparison.
- A chained relational comparison `X op1 Y op2 Z` is evaluated as follows:
  - First, `X op1 Y` is evaluated to `b1`, where the value of `Y`, prior to implicit conversion for overload resolution in `X op1 Y`, is preserved as `y`.
  - If `b1` is `false`, then the evaluation completes and the result is `false`.
  - If `b1` is `true`, evaluate `Z` into `z` and return `y op2 z`.

In other words, `X < Y < Z` can be defined to be equivalent to

```csharp
var x = X;
var y = Y;
if (!(x < y)) return false;
var z = Z;
return y < z;
```

---

Reading the history of `chained-relational-comparison.md`, it appears that:

- The natural, two-**separate**-conversion, semantics had been proposed, in [commit `4200765`](https://github.com/dotnet/csharplang/commit/4200765eabc0c2adadfce7873936e0f68e5e7a6c), with the other rejecting any and all non-identity conversions.
- Later, the proposal was updated to two-**chained**-conversion for the middle operand.

The issue with two-chained-conversion is that it goes against a long-term rule in C# that one expression cannot cause two implicit chained invocations of user-defined implicit conversions. In addition, the chained conversion proposal won't work when `Y` is narrower than `X`, `Z` but `X`, `Z` are "incomparable", e.g., `(double)1 < (int)2 < (decimal)3` makes perfect sense but under the current proposal, it won't compile because `(int)2` is implicitly converted to `double`, which cannot be implicitly converted to `decimal`.

The only downside I can see with the two-separate-conversion is when the middle operand uses target typing (notably lambda expressions), then it's impossible to translate it to `var y = Y;`. In that case, one can consider:

- If `Y` is lambda expression convertible to delegate `D` and in `X op1 Y`, the selected overload is (an extension) `bool operator <(D, D)`, then `Y` will be considered to be of type `D` when resolving `Y op2 Z`.

Or maybe just altogether reject classifying `X op1 Y op2 Z` as chained comparison when `Y` doesn't have a natural type.

**Reply by CyrusNajmabadi** [reply]
@GeeLaw Can you just put the contents here.  So there's no need to go back and forth?

**Reply by CyrusNajmabadi** [reply]
@GeeLaw this part is already flagged in the spec for LDM discussion.

**Reply by CyrusNajmabadi** [reply]
> This note is incorrect

It isn't. The middle value is computed once. The conversions sit on top of that single value and don't recompute it. That's all the note is claiming.

> It could cause the middle operand to be twice-converted by user-defined implicit conversions, which is highly unusual

True.  And agreed on it being unusual. But so is the expression. You're taking one value and comparing it to two different things at once. Folding that down with a conversion on each side isn't that surprising. With user defined conversions especially, I'm more comfortable with one real value flowing through and getting compared to each side than with inventing an intermediate value and gluing two unrelated conversions onto it.

> with Y being its original type (not already subject to implicit conversion during overload resolution of X op1 Y)

This is the part I pushed back on hardest. It asks the right comparison to treat the middle as if the left comparison never touched it, then work it out again from scratch. By the time you reach the right comparison, the left one has already settled what the middle is. Recovering the "before" version means reaching back into something already worked out and pulling it apart. Strange to write down and strange to build.

> ```
> var x = X;
> var y = Y;
 > if (!(x < y)) return false;
 > var z = Z; return y < z;
 > ```

This rewrite doesn't hold up. `var y = Y` assumes `Y` has a type you can capture, and plenty of the time it doesn't. The middle can be a `lambda, null, default, or new()` (and so on and so forth). You can't write `var y = () => 5` or `var y = null`. So the moment you say "grab Y in its original form" you hit operands that have no original form to grab. The model in the spec sidesteps that. The middle gets a settled type from the first comparison and that one value flows into the second. Same value on both sides, evaluated once.

> but under the current proposal, it won't compile

Right, it doesn't, and I'm fine with that. It's an exotic shape that is unlikely to show up in real code. Anyone who wants it can write the two comparisons out.

> Or maybe just altogether reject classifying X op1 Y op2 Z as chained comparison when Y doesn't have a natural type.

Maybe. But that's a narrower, more special cased language than the one in the spec, which works without caring whether the middle has a natural type. That's why I went with it.

All of this is for LDM to land, and it's already flagged in the spec for that.

**Reply by GeeLaw** [reply]
> this part is already flagged in the spec for LDM discussion

In [commit `4200765`](https://github.com/dotnet/csharplang/commit/4200765eabc0c2adadfce7873936e0f68e5e7a6c), your proposal didn't talk about chained conversions at all and **could** be read as two separate conversions. In [commit `192c6ee`](https://github.com/dotnet/csharplang/commit/192c6eea6ce125e542e9a091cd69c4ea10ee80ab), the text becomes that middle is chain-converted.

I would suggest that adding the option of two-separate-conversion into the discussion.

> I don't believe the note is incorrect. 'e_n' itself is only evaluated once.

The note is incorrect not because of number of evaluations, but due to the requirement to spell out more casts. In the example above, `t1 < t3 < t2` **cannot** be rewritten as `t1 < t3 && t3 < t2` (say, when `t1`, `t2`, `t3` are locals, so one-time evaluation doesn't matter). It can only be rewritten as `t1 < t3 && (T1)t3 < t2` (except `t3` is converted to `T1` only once).

> That alone was both extremely strange to have to spec out. But also to implement.

It's not, and even if so, it's not worth to simplify spec to allow weird behaviors. Section 12 of C# 11 extensively talks about types of a subexpression and when resolving `x < y`, the types of `x` and `y` are considered to find the set of user-defined operators (of course, when `y` is a lambda expression, only `x`'s type contributes to the set of user-defined operators).

> The question is: does that matter in practice?

Of course not. One should also consider the question "does having an operand expression in comparison operators that does not have a type matter in practice" (the answer for lambda expression is no to me).

I should be clear about my position:

1. Disallowing `(double)1 < (int)2 < (decimal)3` is completely fine.
2. Disallowing operands that are lambda expressions is completely fine, and highly desired, because there's no natural notion of ordering among functions.
3. Expression from scenario 1 is more likely in practice (all the operands have a natural type, is useful in practice) than expression in scenario 2 (some the operands **in a comparison expression** do not have a natural type, is less useful in practice).
4. Allowing chained user-defined implicit conversions in `t1 < t3 < t2` is problematic. (There's preexisting intuition about no auto-chaining of user-defined conversions. There's also performance/precision issue, because in a realistic scenario, if the middle operand has the narrowest range, then converting directly from the middle type to the two sides is likely more performant and loses less precision if that's a thing.)
5. Allowing separate conversions in `x < y < z` for `y` (so `y` in `y < z` is not subject to `x < y`'s conversion of `y`) is completely fine.

> But the spec has an approach that works for most cases.

Yes, but it enables scenario 2 but not 1, a bad prioritization. My ranking of the approaches:

- The two best are ties, either disallow conversions at all or allow two separate conversions from the natural type of the middle operand.
- Allowing two chained conversions on the middle operand is bad.

> disparate conversions applied to it

They're **not disparate**. Both are "rooted from" middle and converts naturally during overload resolution of the two comparisons.

> but given that hte expr is used twice, once to compare to teh left side, then after that to compare to the right, it's not necessarily strange that then there might be two conversions involved in the construct that collapses things down.

In the two-separate-conversion idea, there also might be two conversions, but there are no chained conversions. Another argument favoring two-separate over two-chained is as follows.

Consider `a < b < c` and assume `operator <` and `operator >` are implemented in sane ways, then rewriting it as `c > b > a` has other effects other than the evaluation order of `a`, `b`, `c`. Namely, in `a < b < c` the expression `b` undergoes conversions `B -> A -> C`, but in `c > b > a` it's `B -> C -> A`. Now consider the two-separate-conversion definition, then in both cases, it's `B -> A` and `B -> C` (order consistent with how things are written).

**Reply by CyrusNajmabadi** [reply]
> In [commit 4200765](https://github.com/dotnet/csharplang/commit/4200765eabc0c2adadfce7873936e0f68e5e7a6c), your proposal didn't talk about chained conversions at all and could be read as two separate conversions. In [commit 192c6ee](https://github.com/dotnet/csharplang/commit/192c6eea6ce125e542e9a091cd69c4ea10ee80ab),

The individual commits are not relevant.  We can discuss the proposal here: https://github.com/dotnet/csharplang/blob/main/proposals/chained-relational-comparison.md

> The note is incorrect 

It's a note.  It's not trying to state exactly what is happening.  It gives a general intuition.   It's what you would tell someone casually (an 'elevator pitch') for what the feature is.  We do this in many of our speclets.  

> Of course not. 

Ok.  Then i don't know why we should care here.  We both agree it's code taht is unlikely to occur in practice.  So trying to make it work isn't important.

> Allowing chained user-defined implicit conversions in t1 < t3 < t2 is problematic.

I *know*. That's why it's an open question for LDM.  It's *very* possible that it is disallowed by LDM.

> It's not

It is. 

> and even if so, it's not worth to simplify spec to allow weird behaviors

My point is that i do not find it weird.  But *again* this will be for LDM to decide.  I can easily see us deciding this simply will not be allowed. 

The spect allows for a very simple, and effective solution.  But it does come with the multiple conversion issue.  If that's a bitter pill, it's trivial to disallow.

**Reply by GeeLaw** [reply]
> why we should care here

I'm not sure if this particular "we" includes me or not. In either case, I care because I am against chained version and I am for both no-conversion and separate-conversion versions. Or in your metaphor, `simple without bitter pill > simple with bitter pill` and `more complex (with a purpose) > simple with bitter pill`.

> This will be for LDM to decide.

If LDM simply requires no-conversion, then that's great. Suppose LDM decides two conversions are acceptable, I would hope LDM to consider two-separate-conversion instead of going directly to two-chained-conversion. Of course, it's your choice to bring the two-separate-conversion idea to LDM or not.

For the *first* time, I would appreciate you simply state whether you would bring the two-separate-conversion idea to LDM, instead of us arguing back and forth about which kind of two conversions is better (I consider the latter issue a personal preference, while the former action could help non-LDT set expectation about what is upcoming).

**Reply by CyrusNajmabadi** [reply]
>  would appreciate you simply state whether you would bring the two-separate-conversion idea to LDM,

Yes. When we bring, we bring all feedback/ideas and whatnot.  And we discuss.

**Reply by GeeLaw** [reply]
Some technical points, about how things can be specced/handled, regarding what you added in-between me typing the reply:

> It asks the right comparison to treat the middle as if the left comparison never touched it, then work it out again from scratch. By the time you reach the right comparison, the left one has already settled what the middle is. Recovering the "before" version means reaching back into something already worked out and pulling it apart.

It's not entirely from scratch. I now have a clearer statement about my proposed idea. It prevents looking at type-to-type implicit conversions (not typeless-to-typed conversions if that's allowed at all) caused by the first comparison for the middle.

> The middle can be a `lambda, null, default, or new()` (and so on and so forth).

Fortunately, `default` and `new()` don't work as of today anyways! Neither `(int)0 < default` nor `(int)0 < new()` compiles in current C#, so the spec naturally disallows them.

For `null`, it can be handled via two-separate-conversion by converting `null` literal separately in the two comparisons. The bonus is that `null` literal conversions are never user-defined, so it's less surprise compared to converting `null` by first comparison, then invoking a potentially user-defined conversion for the second comparison.

I agree lambda/collection expressions are difficult to deal with for two-separate-conversion. For those, the two-chained-conversion appears to be the only choice, if typeless middle and two conversions are allowed.

**Reply by CyrusNajmabadi** [reply]
> Fortunately, default and new() don't work as of today anyways! Neither (int)0 < default nor (int)0 < new() compiles in current C#, so the spec naturally disallows them.

The general concern is simply around untyped expressions.  The spec doesn't blanket disallow them from binary comparison operators. 

> For those, the two-chained-conversion appears to be the only choice,

Agreed.  As such, my preference is that.  As it then handles most of the above cases, and disallows cases no one is fighting for.
